### PR TITLE
Create a wrapper function for bind_Rows so that it can bring in the original data frame name

### PR DESCRIPTION
**Pull request by hidekoji**
# Description

Create a wrapper function for bind_Rows so that it can bring in the original data frame name

# Checklist
Make sure you have performed following items before submitting this pull request.
If not, please describe the reason.  

- [x] Add test cases for this fix/enhancement
- [x] Pass devtools::check()
- [x] Pass devtools::test()
- [ ] Test installing from github
- [x] Tested with Exploratory

### CODE REVIEW

**Comment by kei51e** [comment]
Can one of the admins verify this patch?

**Comment by hideaki** [comment]
Being data wrangling related functions, util.R should be more appropriate place to add this function.

**Comment by hideaki** [comment]
I think converting columns to character once across the board by default is too much.
How about running without that first, and only if errors like below are captured, do it as fallback?
```
> setdiff(data.frame(x = as.character(1:3)), data.frame(x = 1:4))
Error: not compatible: Incompatible type for column `x`: x factor, y integer
> union(data.frame(x = as.character(1:3)), data.frame(x = 1:4))
Error: not compatible: Incompatible type for column `x`: x factor, y integer
> union_all(data.frame(x = as.character(1:3)), data.frame(x = 1:4))
Error: Column `x` can't be converted from factor to integer
> intersect(data.frame(x = as.character(1:3)), data.frame(x = 1:4))
Error: not compatible: Incompatible type for column `x`: x factor, y integer
> bind_rows(data.frame(x = as.character(1:3)), data.frame(x = 1:4))
Error: Column `x` can't be converted from factor to integer
```

**Comment by hidekoji** [comment]
Updated as suggested.

File: R/util.R [hunk 1]
```diff
@@ -1474,3 +1474,167 @@ one_hot <- function(df, key) {
   # Spread the column into multiple columns with name <original column name>_<original value> and value of 1 or 0.
   df %>% tidyr::spread(!!rlang::enquo(key), !!rlang::sym(tmp_value_col), fill = 0, sep = "_") %>% select(-!!rlang::sym(tmp_id_col))
 }
+
+#' Temporary work-around for https://github.com/tidyverse/dplyr/issues/977
```
Comment (hideaki) [inline R/util.R]: I just removed this n_distinct wrapper this morning. Maybe a merge issue? Please make sure not to revive it.

Comment (hidekoji) [reply R/util.R]: Sorry, it seems some merge conflict issue.
I remove the n_distinct.

**Comment by hidekoji** [comment]
Measured performance between `ignore_column_data_type = FALSE` and `ignore_column_data_type = TRUE`

Test Data: US natality (9,153,600 Rows and 23 columns. I added 5 Date columns to see the effect of data type conversion)


###  ignore_column_data_type = FALSE

Expression: microbenchmark(df <- bind_rows(list("a" = usbaby, "b" = usbaby, "c" = usbaby), ignore_column_data_type = FALSE))

Average: 14.39583 (tried 100 times)

###  ignore_column_data_type = TRUE

Expression: microbenchmark(df <- bind_rows(list("a" = usbaby, "b" = usbaby, "c" = usbaby), ignore_column_data_type = TRUE))

Average: 20.32934 (tried 100 times)

So the difference is 6 sec or so.

File: R/util.R [hunk 2]
```diff
@@ -1474,3 +1474,118 @@ one_hot <- function(df, key) {
...
+      new_id <- avoid_conflict(colnames(x[[1]]), .id)
+    } else {
+      new_id <- .id
+    }
+    dplyr::bind_rows(..., .id = new_id)
+  }
+}
+
+#'Wrapper function for dplyr's set operations to support ignoring data type difference.
+set_operation_with_force_caracter <- function(func, x, y, ...) {
```
Comment (hideaki) [inline R/util.R]: Typo: caracter

File: R/util.R [hunk 3]
```diff
@@ -1474,3 +1474,118 @@ one_hot <- function(df, key) {
...
+    } else {
+      new_id  = .id
+    }
+    if(ignore_column_data_type) {
+      readr::type_convert(dplyr::bind_rows(x_updated, .id = new_id))
+    } else {
+      dplyr::bind_rows(x_updated, .id = new_id)
+    }
+  } else {
+    #re-evaluate column data types
```
Comment (hideaki) [inline R/util.R]: Comment does not seem to correctly telling what the code is doing.

File: R/util.R [hunk 4]
```diff
@@ -1474,3 +1474,118 @@ one_hot <- function(df, key) {
...
+
+
+#'Wrapper function for dplyr::bind_rows to support named data frames when it's called inside dplyr chain.
+#'@export
+bind_rows <- function(..., .id = NULL, first_id = '', ignore_column_data_type = TRUE) {
+  # If the dplyr::bind_rows is called within a dplyr chain like df1 %>% dplyr::bind_rows(list(df_2 = df2, df_3 = df3), .id="id"),
+  # since df1 does not have a name, the "id" column of the resulting data frame does not have the data frame name for rows from df1.
+  # To workaround this issue, set a name to the first data frame with the value specified by fistLabel argument as a pre-process
+  # then pass the updated list to dplyr::bind_rows.
+  x_updated <- list()
```
Comment (hideaki) [inline R/util.R]: x_updated, x: better naming, like dfs_updated, dfs.

File: R/util.R [hunk 5]
```diff
@@ -1474,3 +1474,118 @@ one_hot <- function(df, key) {
...
+  x <- dplyr:::flatten_bindable(rlang::dots_values(...))
+  if(ignore_column_data_type | stringr::str_length(first_id) >0) {
+    count <- 0;
+    if(!is.null(names(x))) {
```
Comment (hideaki) [inline R/util.R]: Explain about expected behavior for different ways of passing argument.

File: R/util.R [hunk 6]
```diff
@@ -1474,3 +1474,118 @@ one_hot <- function(df, key) {
...
+      for (name in names(x)) {
+        if(stringr::str_length(first_id) >0 & count == 0) {
```
Comment (hideaki) [inline R/util.R]: Comment on what this case is.

File: R/util.R [hunk 7]
```diff
@@ -1474,3 +1474,118 @@ one_hot <- function(df, key) {
...
+#'@export
+bind_rows <- function(..., .id = NULL, first_id = '', ignore_column_data_type = TRUE) {
+  # If the dplyr::bind_rows is called within a dplyr chain like df1 %>% dplyr::bind_rows(list(df_2 = df2, df_3 = df3), .id="id"),
+  # since df1 does not have a name, the "id" column of the resulting data frame does not have the data frame name for rows from df1.
+  # To workaround this issue, set a name to the first data frame with the value specified by fistLabel argument as a pre-process
+  # then pass the updated list to dplyr::bind_rows.
+  x_updated <- list()
+  x <- dplyr:::flatten_bindable(rlang::dots_values(...))
+  if(ignore_column_data_type | stringr::str_length(first_id) >0) {
+    count <- 0;
```
Comment (hideaki) [inline R/util.R]: Should be called index.

Comment (hideaki) [reply R/util.R]: Start from 1, following R's convention.

File: R/util.R [hunk 8]
```diff
@@ -1474,3 +1474,118 @@ one_hot <- function(df, key) {
...
+          }
+          if(ignore_column_data_type) {
+            x_updated[[name]] <- dplyr::mutate_all(x[[name]], funs(as.character))
+          } else {
+            x_updated[[name]] <- x[[name]]
+          }
+        }
+        count <- count + 1
+      }
+    } else {
```
Comment (hideaki) [inline R/util.R]: Comment on what this case is.

File: R/util.R [hunk 9]
```diff
@@ -1474,3 +1474,118 @@ one_hot <- function(df, key) {
...
+      for(i in 1:length(x)) {
+        # if ignore_column_data_type is set, force character as column data types
+        if(ignore_column_data_type) {
+          x_updated[[i]] <- dplyr::mutate_all(x[[i]], funs(as.character))
+        } else {
+          x_updated[[i]] <- x[[i]]
+        }
+      }
+    }
+    #re-evaluate column data types
```
Comment (hideaki) [inline R/util.R]: Correct comment.

File: R/util.R (hunk 2 above)
Comment (hidekoji) [reply R/util.R]: fixed typo

File: R/util.R (hunk 3 above)
Comment (hidekoji) [reply R/util.R]: fixed comment position.

File: R/util.R (hunk 4 above)
Comment (hidekoji) [reply R/util.R]: changed variable names to dataframes_updated and dataframes

File: R/util.R (hunk 5 above)
Comment (hidekoji) [reply R/util.R]: added comment.

File: R/util.R (hunk 6 above)
Comment (hidekoji) [reply R/util.R]: added comment

File: R/util.R (hunk 7 above)
Comment (hidekoji) [reply R/util.R]: changed it to index and changed it to start from 1

File: R/util.R (hunk 8 above)
Comment (hidekoji) [reply R/util.R]: added comment.

File: R/util.R (hunk 9 above)
Comment (hidekoji) [reply R/util.R]: corrected the comment.

**Pull request merged**